airflow/triggers/base.py

```diff
@@ -135,3 +148,105 @@ def __eq__(self, other):
+
+        # Update the next kwargs of the task instance
+        task_instance.next_kwargs = next_kwargs
+
+        # Remove ourselves as its trigger
+        task_instance.trigger_id = None
+
+        # Set the state of the task instance to scheduled
+        task_instance.state = TaskInstanceState.SCHEDULED
+
+
+class BaseTaskEndEvent(TriggerEvent):
+    """Base event class to end the task without resuming on worker."""
+
+    task_instance_state: TaskInstanceState
+
+    def __init__(self, *, xcoms: dict[str, Any] | None = None, **kwargs) -> None:
+        """
+        Initialize the class with the specified parameters.
+
+        :param xcoms: A dictionary of XComs or None.
+        :param kwargs: Additional keyword arguments.
+        """
+        if "payload" in kwargs:
+            raise ValueError("Param 'payload' not supported for this class.")
+        super().__init__(payload=self.task_instance_state)
+        self.xcoms = xcoms
+
+    @provide_session
+    def handle_submit(self, *, task_instance: TaskInstance, session: Session = NEW_SESSION):
+        """
+        Submit event for the given task instance.
+
+        Marks the task with the state `task_instance_state` and optionally pushes xcom if applicable.
+
+        :param task_instance: The task instance to be submitted.
+        :param session: The session to be used for the database callback sink.
+        """
+        # Mark the task with terminal state and prevent it from resuming on worker
+        task_instance.trigger_id = None
+        task_instance.state = self.task_instance_state
+
+        self._submit_callback_if_necessary(task_instance=task_instance, session=session)
+        self._push_xcoms_if_necessary(task_instance=task_instance)
+
+    def _submit_callback_if_necessary(self, *, task_instance: TaskInstance, session):
+        """Submit a callback request if the task state is SUCCESS or FAILED."""
+        is_failure = self.task_instance_state == TaskInstanceState.FAILED
+        if self.task_instance_state in [TaskInstanceState.SUCCESS, TaskInstanceState.FAILED]:
+            request = TaskCallbackRequest(
+                full_filepath=task_instance.dag_model.fileloc,
+                simple_task_instance=SimpleTaskInstance.from_ti(task_instance),
+                is_failure_callback=is_failure,
+                task_callback_type=self.task_instance_state,
+            )
+            log.warning("Sending callback: %s", request)
+            try:
+                DatabaseCallbackSink().send(callback=request, session=session)
+            except Exception as e:
+                log.error("Failed to send callback: %s", e)
+
+    def _push_xcoms_if_necessary(self, *, task_instance: TaskInstance):
+        """Pushes XComs to the database if they are provided."""
+        if self.xcoms:
+            for key, value in self.xcoms.items():
+                task_instance.xcom_push(key=key, value=value)
+
+
+class TaskSuccessEvent(BaseTaskEndEvent):
```

Comment:
This interface now will be used by the provider right do we need to document it? also, is there is way to make it experimental, I'm assuming this will be an experimental feature

Reply:
Yes I will add the documentation